### PR TITLE
Close #270: Permanent program addresses

### LINKED ISSUE #270: Permanent program addresses
Currently programs have addresses built from hash of concatenation of hash of first version of program and transaction id. @Chicker proposed idea of permanent program addresses. It will be convenient when we deploy must-have programs (such as GT) on TestNet after reset.

### Changes

1. User defines address of a program.
2. To create or update a program user should show a signature of code owned by defined address.
3. Now programs has no owner.

### Signing policy

1. To _create a program_ user should sign hash of code of new program
2. To _update program_ user should sign hash of concatenation of old code and new code.
3. To _seal a program_ user should sing hash of concatenation of `Seal` word (in UTF8 encoding) and stored code of a program.

### Cryptography

|Signatures|Hash|
|-------|------|
|[ed25519](https://ed25519.cr.yp.to)|SHA-512|

### PR DESCRIPTION
**Pull request by fomkin**

### CODE REVIEW

File: vm/src/main/scala/pravda/vm/operations/SystemOperations.scala [hunk 1]
```diff
@@ -101,51 +119,75 @@ final class SystemOperations(program: ByteBuffer,
...
+      case Some(ProgramContext(_, _, true)) => throw ThrowableVmError(Error.ProgramIsSealed)
+      case None                             => throw ThrowableVmError(Error.NoSuchProgram)
     }
   }
 
   @OpcodeImplementation(
     opcode = PCREATE,
-    description = "Takes bytecode of a new program, put's it to " +
-      "state and returns program address."
+    description = "Takes address (pubKey), bytecode, and it's ed25519 signature. " +
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/operations/SystemOperations.scala]:
```suggestion
    description = "Takes address (pubKey), bytecode, and its ed25519 signature. " +
```

File: vm/src/main/scala/pravda/vm/operations/SystemOperations.scala [hunk 2]
```diff
@@ -101,51 +119,75 @@ final class SystemOperations(program: ByteBuffer,
...
+      "If signature is invalid and program on the address is not existed before create new program."
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/operations/SystemOperations.scala]:
```suggestion
      "If signature is valid and program didn't exist before on the specified address create new program."
```

File: vm/src/main/scala/pravda/vm/operations/SystemOperations.scala [hunk 3]
```diff
@@ -101,51 +119,75 @@ final class SystemOperations(program: ByteBuffer,
 
   @OpcodeImplementation(
     opcode = PUPDATE,
-    description = "Takes address of a program and new bytecode. " +
-      "Replaces bytecode in storage. This opcode can be performed " +
-      "only from owner of the program"
+    description = "Takes address of an existing program, code, and signature. " +
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/operations/SystemOperations.scala]:
```suggestion
    description = "Takes address of an existing program, code and signature. " +
```

File: vm/src/main/scala/pravda/vm/operations/SystemOperations.scala [hunk 4]
```diff
@@ -101,51 +119,75 @@ final class SystemOperations(program: ByteBuffer,
...
+      "Message for signing is result of concatenation of old code new code of the program"
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/operations/SystemOperations.scala]:
```suggestion
      "Signature is computed from concatenation of old code and new code of the program"
```

File: vm/src/main/scala/pravda/vm/operations/SystemOperations.scala [hunk 5]
```diff
@@ -101,51 +119,75 @@ final class SystemOperations(program: ByteBuffer,
...
+        env.createProgram(programAddress, code)
+      case _ =>
+        throw ThrowableVmError(Error.OperationDenied)
+    }
   }
 
   @OpcodeImplementation(
     opcode = SEAL,
-    description = "Takes the address of an existing program, makes the program sealed"
+    description = "Takes the address of an existing program, and signature of code with seal mark. " +
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/operations/SystemOperations.scala]:
```suggestion
    description = "Takes the address of an existing program and signature of code with seal mark. " +
```

File: vm/src/main/scala/pravda/vm/operations/SystemOperations.scala [hunk 6]
```diff
@@ -101,51 +119,75 @@ final class SystemOperations(program: ByteBuffer,
...
+      "Message for signing is result of concatenation of 'Seal' word (in UTF8 encoding) and current code of program."
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/operations/SystemOperations.scala]:
```suggestion
      "Signature is computed from concatenation of 'Seal' word (in UTF8 encoding) and current code of program."
```

File: vm/src/main/scala/pravda/vm/operations/SystemOperations.scala [hunk 7]
```diff
@@ -248,3 +278,7 @@ final class SystemOperations(program: ByteBuffer,
     }
   }
 }
+
+object SystemOperations {
+  final val SealTag = ByteString.copyFromUtf8("sealed")
```
Comment (vovapolu) [inline vm/src/main/scala/pravda/vm/operations/SystemOperations.scala]: `"sealed"` or `"Seal"`?

File: vm/src/test/scala/pravda/vm/VmSandbox.scala [hunk 8]
```diff
@@ -15,32 +14,25 @@ object VmSandbox {
                            pExecutor: Address)
       extends Environment {
 
-    val balances = mutable.Map(initBalances: _*)
-    val programs = mutable.Map(initPrograms: _*)
-    val sealedPrograms = mutable.Map[Address, Boolean]()
+    private val balances = mutable.Map(initBalances: _*)
```
Comment (vovapolu) [inline vm/src/test/scala/pravda/vm/VmSandbox.scala]: why?

File: vm/src/main/scala/pravda/vm/operations/SystemOperations.scala (hunk 7 above)
Comment (fomkin) [reply vm/src/main/scala/pravda/vm/operations/SystemOperations.scala]: Does it matter?

File: vm/src/test/scala/pravda/vm/VmSandbox.scala (hunk 8 above)
Comment (fomkin) [reply vm/src/test/scala/pravda/vm/VmSandbox.scala]: Why not?

File: vm/src/main/scala/pravda/vm/operations/SystemOperations.scala (hunk 7 above)
Comment (vovapolu) [reply vm/src/main/scala/pravda/vm/operations/SystemOperations.scala]: In docs you write that it's "Seal".

File: vm/src/test/scala/pravda/vm/VmSandbox.scala (hunk 8 above)
Comment (vovapolu) [reply vm/src/test/scala/pravda/vm/VmSandbox.scala]: It's possible that in future we will need to access these fields.

Comment (fomkin) [reply vm/src/test/scala/pravda/vm/VmSandbox.scala]: It's a bad practice to expose mutable state to public.

Comment (fomkin) [reply vm/src/test/scala/pravda/vm/VmSandbox.scala]: If we will need to access this data we should add a method.

File: vm/src/main/scala/pravda/vm/operations/SystemOperations.scala (hunk 7 above)
Comment (fomkin) [reply vm/src/main/scala/pravda/vm/operations/SystemOperations.scala]: Ok, it's true.

**Pull request merged**